### DISCUSSION TITLE
CLI target commands

### DISCUSSION THREAD

**Discussion by PaulFarault** [TOSIT-IO/TDP#91] [Ideas]
Even if it was not originally planned, the tdp-lib CLI has become the reference CLI for administering a TDP cluster. New commands were added gradually without planning. This has led to some inconsistencies.

I'm opening this discussion to try to establish a target for the next release.

Here are the commands currently available:

```txt
browse        Browse deployments.
dag           Compute and display a graph of the DAG.
default-diff  Difference between tdp_vars and defaults.
deploy        Execute a planned deployment.
init          Initialize the database and the TDP variables.
ops           Display all available operations.
plan          Generate a deployment plan.
    edit             Edit the planned deployment.
    dag              Deploy from the DAG. 
    ops              Run a list of operations.
    resume           Resume a failed deployment.
    reconfigure      Reconfigure required TDP services.
playbooks     Generate meta playbooks in order to use a TDP like...
status        Manage the status of the cluster.
    edit                Edit the status of the cluster.
    generate-stales     Generate stale components.
    show                Print the status of the cluster.
validate      Validate TDP variables against the loaded collections...
vars          Manage services and components configurations.
    edit                  Edit a variables file.
```

Also, here are some features that we'd like to add:

- Display variables that apply for a service/component
- Display the deployment history of a service/component
- Import a deployment plan from a file
- Import variables from a file

I may miss some wanted features.

---

### Proposal

I suggest the following commands / sub-commands for the CLI. I don't include all available options, only the one that represent a new feature or a current command that should be refactored:

```txt
browse                   Browse past deployments logs.
deploy                   Execute a planned deployment.
init                     Initialize the database and the variables.
ops                      Display all available operations.
    --graph                **REFACTORED** Compute and display a graph of DAG operations.
plan                     Plan a deployment.
    edit                   Edit the planned deployment.
    create                 **NEW** Plan a new deployment.
        dag                  **REFACTORED** Plan a new deployment based on the DAG.
        ops                  **REFACTORED** Plan a new deployment from a list of operations.
        resume               **REFACTORED** Plan a new deployment from a failed one.
        reconfigure          **REFACTORED** Plan a new deployment from the list of stale components.
    show                   **NEW** Display the current planned deployment.
    import                 **NEW** Import a deployment plan from a file.
status                   Browse and manage the status of the cluster.
    edit                   Override the status of the cluster.
    generate-stales        Update stale components status.
    show                   Display the cluster status.
        --history            **NEW** Display individual modifications.
vars                     Manage services and components configurations.
    edit                   Edit a variables file.
    validate               **REFACTORED** Validate variables against the provided schema.
    show                   **NEW** Display variables that apply for a given service or component.
    import                 **NEW** Import variables from a file.
```

I removed `default-diff` and `playbooks` as they should be converted to a script IMO.

I'm open to any suggestions.

**Comment by giem-git** [comment]
Hey, Paul. Your proposal makes sense. 

The `plan create` option seems to me a little cumbersome. Maybe this would be simpler : 
  ~~~
  plan
    edit
    from-dag
    from-ops
    resume
    reconfigure
    show
    import
  ~~~
Just a personal view, keep up the good work :)

**Reply by giem-git** [reply]
Or just `from` instead of `create`. I don't like `create resume`, the following seems to me closer to natural language :
```
plan from dag
plan from ops
plan from failed
plan from stales
```

**Reply by gonzaloetjo** [reply]
Agreed. I'd add that, `plan import` could be moved to `plan from file` or `plan from import`, as it targets the same goal: creating a new deployment plan.

Also, it would be interesting if `show` could be implemented as a flag `--show`. In order to allow visualization of potential new plans being implemented, something like `plan from failed --show`.

**Reply by PaulFarault** [reply]
I agree with @gonzaloetjo , to gather all commands that generate a new deployment plan in the same group.

`tdp plan show` is made to display the existing plan (if any). If you want to preview a plan without saving it into the database you can already use `tdp plan from xxx --preview`. Is that what you meant or did I misunderstood your use case?
